src/ripple/app/misc/impl/ValidatorList.cpp

```diff
@@ -229,152 +299,653 @@ ValidatorList::CacheValidatorFile(
+    auto result = splitMessageParts(messages, largeMsg, maxSize, begin, mid);
+    return result + splitMessageParts(messages, largeMsg, maxSize, mid, end);
+}
+
+std::size_t
+splitMessageParts(
+    std::vector<ValidatorList::MessageWithHash>& messages,
+    protocol::TMValidatorListCollection const& largeMsg,
+    std::size_t maxSize,
+    std::size_t begin,
+    std::size_t end)
+{
+    if (end <= begin)
+        return 0;
+    if (end - begin == 1)
+    {
+        protocol::TMValidatorList smallMsg;
+        smallMsg.set_version(1);
+        smallMsg.set_manifest(largeMsg.manifest());
+
+        auto const& blob = largeMsg.blobs(begin);
+        smallMsg.set_blob(blob.blob());
+        smallMsg.set_signature(blob.signature());
+        // This is only possible if "downgrading" a v2 UNL to v1.
+        if (blob.has_manifest())
+            smallMsg.set_manifest(blob.manifest());
+
+        auto message =
+            std::make_shared<Message>(smallMsg, protocol::mtVALIDATORLIST);
+
+        if (message->getBuffer(compression::Compressed::Off).size() <= maxSize)
+        {
+            messages.emplace_back(message, sha512Half(smallMsg), 1);
+            return messages.back().numVLs;
+        }
+    }
+    else
+    {
+        boost::optional<protocol::TMValidatorListCollection> smallMsg;
+        smallMsg.emplace();
+        smallMsg->set_version(largeMsg.version());
+        smallMsg->set_manifest(largeMsg.manifest());
+
+        for (std::size_t i = begin; i < end; ++i)
+        {
+            *smallMsg->add_blobs() = largeMsg.blobs(i);
+        }
+        auto message = std::make_shared<Message>(
+            *smallMsg, protocol::mtVALIDATORLISTCOLLECTION);
+
+        if (message->getBuffer(compression::Compressed::Off).size() > maxSize)
+        {
+            // free up the message space
+            message.reset();
+            smallMsg.reset();
+            return splitMessage(messages, largeMsg, maxSize, begin, end);
+        }
+        else
+        {
+            messages.emplace_back(
+                message, sha512Half(*smallMsg), smallMsg->blobs_size());
+            return messages.back().numVLs;
+        }
+    }
+    return 0;
+}
+
+// Build a v1 protocol message using only the current VL
+std::size_t
+buildValidatorListMessage(
+    std::vector<ValidatorList::MessageWithHash>& messages,
+    std::uint32_t rawVersion,
+    std::string const& rawManifest,
+    ValidatorBlobInfo const& currentBlob,
+    std::size_t maxSize)
+{
+    assert(messages.empty());
+    protocol::TMValidatorList msg;
+    auto const manifest =
+        currentBlob.manifest ? *currentBlob.manifest : rawManifest;
+    auto const version = 1;
+    msg.set_manifest(manifest);
+    msg.set_blob(currentBlob.blob);
+    msg.set_signature(currentBlob.signature);
+    // Override the version
+    msg.set_version(version);
+
+    auto message = std::make_shared<Message>(msg, protocol::mtVALIDATORLIST);
```

Comment:
See [this comment](https://github.com/ripple/rippled/pull/3619/files#r507995736).